File: 2342.max-sum-of-a-pair-with-equal-sum-of-digits.py
#
# @lc app=leetcode id=2342 lang=python3
#
# [2342] Max Sum of a Pair With Equal Sum of Digits
#

# @lc code=start
from collections import Counter
class Solution:
    def maximumSum(self, nums: list[int]) -> int:
        hashmap = Counter()
        ans = 0
        for num in nums:
            sum_digits = sum(map(int, str(num)))
            #Update the answer
            if sum_digits in hashmap:
                ans = max(ans, hashmap[sum_digits] + num)
            #Update the hashmap
            hashmap[sum_digits] = max(hashmap[sum_digits], num)
        return -1 if ans == 0 else ans
# @lc code=end

File: 2342.max-sum-of-a-pair-with-equal-sum-of-digits.java
/*
 * @lc app=leetcode id=2342 lang=java
 *
 * [2342] Max Sum of a Pair With Equal Sum of Digits
 */

// @lc code=start
import java.util.HashMap;
import java.util.Map;


class Solution {
    public int maximumSum(int[] nums) {
        Map<Integer, Integer> map = new HashMap<>();
        int ans = -1;
        for (int num: nums){
            int sum = digitSum(num);
            if (map.containsKey(sum)){
                ans = Math.max(ans, map.get(sum) + num);
            }
            map.merge(sum, num, Integer::max);
        }
        return ans;
    }


    private int digitSum(int n){
        int sum = 0;
        while (n > 0) {
            sum += n % 10;
            n /= 10;
        }
        return sum;
    }
}
// @lc code=end

